C error and SIGSEGV on aliased array types comparison

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.ee3a182
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr
vexe: /home/ge/.vlang/v
vexe mtime: 2024-11-13 21:01:15

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.1
Git vroot status: weekly.2024.46-18-gee3a182d
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg segv.v`

```v
module main

type Array1 = []u8

type Array2 = []u8

fn (a Array2) == (other Array2) bool {
	return a == other
}

fn (a Array2) < (other Array2) bool {
	return a < other
}

fn main() {
	a := Array1([u8(127), 0, 0, 1])
	b := Array1([u8(127), 0, 0, 1])

	c := Array2([u8(127), 0, 0, 1])
	d := Array2([u8(127), 0, 0, 1])

	println(a == b) // true
	println(a < b) // C error: invalid operand types for binary operation

	println(c == d) // compiles, but SIGSEGV at runtime
	println(c < d) // compiles, but SIGSEGV at runtime
}
```

**What did you expect to see?**

V checker errors

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/segv.01JCKNPRZSTJ196NY4V0JTWFD4.tmp.c:13746: error: invalid operand types for binary operation
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzM1MTY0N2U3Y2M1YTc4NTQyYzM0MjIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.EVGdNRrG7hmaGku18DrVqYUEWg682Nc5ZpuHF323oxQ">Huly&reg;: <b>V_0.6-21296</b></a></sub>
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -516,8 +516,14 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 		.gt, .lt, .ge, .le {
 			unwrapped_left_type := c.unwrap_generic(left_type)
 			left_sym = c.table.sym(unwrapped_left_type)
+			if left_sym.kind == .alias && !left_sym.has_method_with_generic_parent(node.op.str()) {
+				left_sym = c.table.final_sym(unwrapped_left_type)
+			}
 			unwrapped_right_type := c.unwrap_generic(right_type)
 			right_sym = c.table.sym(unwrapped_right_type)
+			if right_sym.kind == .alias && !right_sym.has_method_with_generic_parent(node.op.str()) {
+				right_sym = c.table.final_sym(unwrapped_right_type)
+			}
 			if left_sym.kind in [.array, .array_fixed] && right_sym.kind in [.array, .array_fixed] {
 				c.error('only `==` and `!=` are defined on arrays', node.pos)
 			} else if left_sym.kind == .struct

--- a/vlib/v/checker/tests/alias_array_unknown_op_overloading_err.out
+++ b/vlib/v/checker/tests/alias_array_unknown_op_overloading_err.out
@@ -4,10 +4,38 @@ vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:18:4: error: unde
    18 |     a -= b
       |       ~~
    19 |     println(a - b)
-   20 | }
+   20 |     println(a < b)
 vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:19:10: error: undefined operation `Tuple` - `Tuple`
    17 |     b := new_tuple(12, 4.5, 6.7, 6)
    18 |     a -= b
    19 |     println(a - b)
       |             ~~~~~
-   20 | }
+   20 |     println(a < b)
+   21 |     println(a > b)
+vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:20:12: error: only `==` and `!=` are defined on arrays
+   18 |     a -= b
+   19 |     println(a - b)
+   20 |     println(a < b)
+      |               ^
+   21 |     println(a > b)
+   22 |     println(a <= b)
+vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:21:12: error: only `==` and `!=` are defined on arrays
+   19 |     println(a - b)
+   20 |     println(a < b)
+   21 |     println(a > b)
+      |               ^
+   22 |     println(a <= b)
+   23 |     println(a >= b)
+vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:22:12: error: only `==` and `!=` are defined on arrays
+   20 |     println(a < b)
+   21 |     println(a > b)
+   22 |     println(a <= b)
+      |               ~~
+   23 |     println(a >= b)
+   24 | }
+vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv:23:12: error: only `==` and `!=` are defined on arrays
+   21 |     println(a > b)
+   22 |     println(a <= b)
+   23 |     println(a >= b)
+      |               ~~
+   24 | }

--- a/vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv
+++ b/vlib/v/checker/tests/alias_array_unknown_op_overloading_err.vv
@@ -17,4 +17,8 @@ fn main() {
 	b := new_tuple(12, 4.5, 6.7, 6)
 	a -= b
 	println(a - b)
+	println(a < b)
+	println(a > b)
+	println(a <= b)
+	println(a >= b)
 }

--- a/vlib/v/tests/builtin_arrays/array_of_alias_op_overload_test.v
+++ b/vlib/v/tests/builtin_arrays/array_of_alias_op_overload_test.v
@@ -1,0 +1,31 @@
+type Array1 = []u8
+
+type Array2 = []u8
+
+fn (a Array2) == (other Array2) bool {
+	return a[0] == other[0]
+}
+
+fn (a Array2) < (other Array2) bool {
+	return a[0] < other[0]
+}
+
+fn test_array_of_alias_op_overload() {
+	a := Array1([u8(127), 0, 0, 1])
+	b := Array1([u8(127), 0, 0, 1])
+
+	c := Array2([u8(127), 0, 0, 1])
+	d := Array2([u8(127), 0, 0, 1])
+
+	ret0 := a == b
+	println(ret0)
+	assert ret0
+
+	ret1 := c == d
+	println(ret1)
+	assert ret1
+
+	ret2 := c < d
+	println(ret2)
+	assert !ret2
+}

